drivers/sensor/bma4xx/bma4xx.c

```diff
@@ -0,0 +1,687 @@
+				  const struct sensor_value *val)
+{
+	struct bma4xx_data *bma4xx = dev->data;
+	uint8_t reg_addr;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		reg_addr = BMA4XX_REG_OFFSET_0;
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		reg_addr = BMA4XX_REG_OFFSET_1;
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		reg_addr = BMA4XX_REG_OFFSET_2;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	/* micro-Gs */
+	int32_t ug = sensor_ms2_to_ug(val);
+
+	if (ug < BMA4XX_OFFSET_MICROG_MIN || ug > BMA4XX_OFFSET_MICROG_MAX) {
+		return -ERANGE;
+	}
+
+	int8_t reg_val = ug / BMA4XX_OFFSET_MICROG_PER_BIT;
+
+	return bma4xx->hw_ops->write_reg(dev, reg_addr, reg_val);
+}
+
+static const uint32_t odr_to_reg_map[] = {
+	0,          /* Invalid */
+	781250,     /* 0.78125 Hz (25/32) => 0x1 */
+	1562500,    /* 1.5625 Hz (25/16) => 0x2 */
+	3125000,    /* 3.125 Hz (25/8) => 0x3 */
+	6250000,    /* 6.25 Hz (25/4) => 0x4 */
+	12500000,   /* 12.5 Hz (25/2) => 0x5 */
+	25000000,   /* 25 Hz => 0x6 */
+	50000000,   /* 50 Hz => 0x7*/
+	100000000,  /* 100 Hz => 0x8*/
+	200000000,  /* 200 Hz => 0x9*/
+	400000000,  /* 400 Hz => 0xa*/
+	800000000,  /* 800 Hz => 0xb*/
+	1600000000, /* 1600 Hz => 0xc*/
+};
+
+static int bma4xx_odr_to_reg(uint32_t microhertz, uint8_t *reg_val)
+{
+	if (microhertz == 0) {
+		/* Illegal ODR value */
+		return -ERANGE;
+	}
+
+	for (uint8_t i = 0; i < ARRAY_SIZE(odr_to_reg_map); i++) {
+		if (microhertz <= odr_to_reg_map[i]) {
+			*reg_val = i;
+			return 0;
+		}
+	}
+
+	/* Requested ODR too high */
+	return -ERANGE;
+}
+
+/**
+ * Set the sensor's acceleration offset (per axis). Use bma4xx_commit_nvm() to save these
+ * offsets to nonvolatile memory so they are automatically set during power-on-reset.
+ */
+static int bma4xx_attr_set_odr(const struct device *dev, const struct sensor_value *val)
+{
+	int status;
+	uint8_t reg_val;
+
+	/* Convert ODR Hz value to microhertz and round up to closest register setting */
+	status = bma4xx_odr_to_reg(val->val1 * 1000000 + val->val2, &reg_val);
+	if (status < 0) {
+		return status;
+	}
+
+	return bma4xx_reg_update(dev, BMA4XX_REG_ACCEL_CONFIG, BMA4XX_MASK_ACC_CONF_ODR, reg_val);
+}
+
+static const uint32_t fs_to_reg_map[] = {
+	2,  /* +/-2G => 0x0 */
+	4,  /* +/-4G => 0x1 */
+	8,  /* +/-8G => 0x2 */
+	16, /* +/-16G => 0x3 */
+};
+
+static int bma4xx_fs_to_reg(uint32_t range_g, uint8_t *reg_val)
+{
+	if (range_g == 0) {
+		/* Illegal value */
+		return -ERANGE;
+	}
+
+	for (uint8_t i = 0; i < ARRAY_SIZE(fs_to_reg_map); i++) {
+		if (range_g <= fs_to_reg_map[i]) {
+			*reg_val = i;
+			return 0;
+		}
+	}
+
+	/* Requested range too high */
+	return -ERANGE;
+}
+
+/**
+ * Set the sensor's full-scale range
+ */
+static int bma4xx_attr_set_range(const struct device *dev, const struct sensor_value *val)
+{
+	struct bma4xx_data *bma4xx = dev->data;
+	int status;
+	uint8_t reg_val;
+
+	/* Convert m/s^2 to G's and find closest register setting */
+	status = bma4xx_fs_to_reg(sensor_ms2_to_g(val), &reg_val);
```

Comment:
I think you need to use micro-g here. If someone sets it to 2.5g (24.516625m/s2) then `sensor_ms2_to_g` will give you 2 and you'll have a range that's too small for what you needed.

Reply:
Will review math tomorrow.

Reply:
You're right this would round down to integer Gs which is no good.